File: Rust/tests/unit/test_checksum.rs
use wip_rust::wip_common_rs::packet::core::checksum::{calc_checksum12, verify_checksum12, embed_checksum12_le};

mod checksum_tests {
    use super::*;

    #[test]
    fn test_calc_checksum12_basic() {
        let data = vec![0x12, 0x34, 0x56, 0x78];
        let checksum = calc_checksum12(&data);
        
        // Checksum should be within 12-bit range (0-4095)
        assert!(checksum <= 0xFFF);
    }

    #[test]
    fn test_calc_checksum12_empty() {
        let data = vec![];
        let checksum = calc_checksum12(&data);
        
        // Empty data should still produce valid checksum
        assert!(checksum <= 0xFFF);
    }

    #[test]
    fn test_calc_checksum12_deterministic() {
        let data = vec![0x12, 0x34, 0x56, 0x78];
        let checksum1 = calc_checksum12(&data);
        let checksum2 = calc_checksum12(&data);
        
        // Same data should produce same checksum
        assert_eq!(checksum1, checksum2);
    }

    #[test]
    fn test_calc_checksum12_different_data() {
        let data1 = vec![0x12, 0x34, 0x56, 0x78];
        let data2 = vec![0x87, 0x65, 0x43, 0x21];
        
        let checksum1 = calc_checksum12(&data1);
        let checksum2 = calc_checksum12(&data2);
        
        // Different data should (likely) produce different checksums
        // Note: This is probabilistic, but very likely to be true
        assert_ne!(checksum1, checksum2);
    }

    #[test]
    fn test_embed_and_verify_checksum12() {
        let mut data = vec![0x12, 0x34, 0x56, 0x78, 0x00, 0x00]; // Last 2 bytes for checksum
        
        // Embed checksum
        embed_checksum12_le(&mut data);
        
        // Verify checksum
        assert!(verify_checksum12(&data).is_ok());
    }

    #[test]
    fn test_verify_checksum12_invalid() {
        let mut data = vec![0x12, 0x34, 0x56, 0x78, 0x00, 0x00];
        
        // Embed valid checksum
        embed_checksum12_le(&mut data);
        
        // Corrupt the data
        data[0] = 0xFF;
        
        // Verify should fail
        assert!(verify_checksum12(&data).is_err());
    }

    #[test]
    fn test_verify_checksum12_short_data() {
        let data = vec![0x12]; // Too short for checksum
        
        // Should handle gracefully
        let result = verify_checksum12(&data);
        assert!(result.is_err());
    }

    #[test]
    fn test_checksum12_large_data() {
        let mut data = vec![0xAA; 1000]; // 1KB of data
        data.extend_from_slice(&[0x00, 0x00]); // Space for checksum
        
        embed_checksum12_le(&mut data);
        assert!(verify_checksum12(&data).is_ok());
    }

    #[test]
    fn test_checksum12_edge_cases() {
        // Test with all zeros
        let mut data = vec![0x00; 10];
        embed_checksum12_le(&mut data);
        assert!(verify_checksum12(&data).is_ok());
        
        // Test with all 0xFF
        let mut data = vec![0xFF; 8];
        data.extend_from_slice(&[0x00, 0x00]);
        embed_checksum12_le(&mut data);
        assert!(verify_checksum12(&data).is_ok());
    }

    #[test]
    fn test_checksum12_known_values() {
        // Test with known input/output for regression testing
        let test_cases = vec![
            (vec![0x01, 0x02, 0x00, 0x00], true),  // Simple case
            (vec![0xFF, 0xFF, 0x00, 0x00], true),  // High values
            (vec![0x00, 0x00, 0x00, 0x00], true),  // All zeros
        ];
        
        for (mut data, should_pass) in test_cases {
            embed_checksum12_le(&mut data);
            let result = verify_checksum12(&data);
            assert_eq!(result.is_ok(), should_pass);
        }
    }

    #[test]
    fn test_checksum12_byte_order() {
        // Test that little-endian embedding works correctly
        let mut data = vec![0x12, 0x34, 0x00, 0x00];
        embed_checksum12_le(&mut data);
        
        // The checksum should be embedded in the last 2 bytes in little-endian format
        // We don't test specific values since the checksum algorithm may vary,
        // but we ensure it validates correctly
        assert!(verify_checksum12(&data).is_ok());
        
        // Test that corrupting the checksum bytes fails validation
        data[data.len() - 1] ^= 0x01; // Flip one bit
        assert!(verify_checksum12(&data).is_err());
    }

    #[test]
    fn test_checksum12_multiple_operations() {
        let mut data = vec![0x12, 0x34, 0x56, 0x78, 0x9A, 0xBC, 0x00, 0x00];
        
        // Embed checksum multiple times - should be idempotent when done correctly
        embed_checksum12_le(&mut data);
        let first_checksum = data.clone();
        
        // Clear checksum area and re-embed
        let data_len = data.len();
        data[data_len - 2] = 0x00;
        data[data_len - 1] = 0x00;
        embed_checksum12_le(&mut data);
        
        assert_eq!(data, first_checksum);
        assert!(verify_checksum12(&data).is_ok());
    }
}

File: cpp/tests/unit/test_checksum.cpp
#include <gtest/gtest.h>
#include "wiplib/packet/checksum.hpp"

using namespace wiplib::packet;

class ChecksumTest : public ::testing::Test {
protected:
    void SetUp() override {}
    void TearDown() override {}
};

// 基本的なチェックサム計算テスト
TEST_F(ChecksumTest, BasicChecksumCalculation) {
    std::vector<uint8_t> data = {0x01, 0x02, 0x03, 0x04};
    uint16_t checksum = calc_checksum12(data);
    
    // 12ビットチェックサムは0-4095の範囲
    EXPECT_LE(checksum, 0x0FFF);
}

// 空データのテスト
TEST_F(ChecksumTest, EmptyDataChecksum) {
    std::vector<uint8_t> empty_data;
    uint16_t checksum = calc_checksum12(empty_data);
    EXPECT_EQ(checksum, 0);
}

// 単一バイトのテスト
TEST_F(ChecksumTest, SingleByteChecksum) {
    std::vector<uint8_t> data = {0xFF};
    uint16_t checksum = calc_checksum12(data);
    EXPECT_LE(checksum, 0x0FFF);
    EXPECT_GT(checksum, 0);
}

// チェックサム検証テスト
TEST_F(ChecksumTest, ChecksumVerification) {
    std::vector<uint8_t> data = {0x12, 0x34, 0x56, 0x78};
    uint16_t checksum = calc_checksum12(data);
    
    // 正しいチェックサムでの検証
    EXPECT_TRUE(verify_checksum12(data, checksum));
    
    // 間違ったチェックサムでの検証
    EXPECT_FALSE(verify_checksum12(data, checksum + 1));
}

// 大きなデータでのテスト
TEST_F(ChecksumTest, LargeDataChecksum) {
    std::vector<uint8_t> large_data(1024);
    for (size_t i = 0; i < large_data.size(); ++i) {
        large_data[i] = static_cast<uint8_t>(i & 0xFF);
    }
    
    uint16_t checksum = calc_checksum12(large_data);
    EXPECT_LE(checksum, 0x0FFF);
    EXPECT_TRUE(verify_checksum12(large_data, checksum));
}

// キャリーフォールドのテスト
TEST_F(ChecksumTest, CarryFoldHandling) {
    // キャリーが発生するデータパターン
    std::vector<uint8_t> data = {0xFF, 0xFF, 0xFF, 0xFF};
    uint16_t checksum = calc_checksum12(data);
    
    EXPECT_LE(checksum, 0x0FFF);
    EXPECT_TRUE(verify_checksum12(data, checksum));
}

// 境界値テスト
TEST_F(ChecksumTest, BoundaryValues) {
    // 最小値
    std::vector<uint8_t> min_data = {0x00};
    uint16_t min_checksum = calc_checksum12(min_data);
    EXPECT_EQ(min_checksum, 0);
    
    // 最大値
    std::vector<uint8_t> max_data = {0xFF};
    uint16_t max_checksum = calc_checksum12(max_data);
    EXPECT_LE(max_checksum, 0x0FFF);
}

// 同じデータは同じチェックサムを生成するテスト
TEST_F(ChecksumTest, Deterministic) {
    std::vector<uint8_t> data = {0xAB, 0xCD, 0xEF};
    uint16_t checksum1 = calc_checksum12(data);
    uint16_t checksum2 = calc_checksum12(data);
    
    EXPECT_EQ(checksum1, checksum2);
}

// 異なるデータは異なるチェックサムを生成する可能性が高いテスト
TEST_F(ChecksumTest, DifferentDataDifferentChecksum) {
    std::vector<uint8_t> data1 = {0x01, 0x02, 0x03};
    std::vector<uint8_t> data2 = {0x03, 0x02, 0x01};
    
    uint16_t checksum1 = calc_checksum12(data1);
    uint16_t checksum2 = calc_checksum12(data2);
    
    // 必ずしも異なるとは限らないが、通常は異なるはず
    EXPECT_NE(checksum1, checksum2);
}